crypto/asn1/d2i_pr.c

```diff
@@ -15,15 +15,70 @@
 #include <openssl/bn.h>
 #include <openssl/evp.h>
 #include <openssl/objects.h>
+#include <openssl/decoder.h>
 #include <openssl/engine.h>
 #include <openssl/x509.h>
 #include <openssl/asn1.h>
 #include "crypto/asn1.h"
 #include "crypto/evp.h"
+#include "internal/asn1.h"
 
-EVP_PKEY *d2i_PrivateKey_ex(int type, EVP_PKEY **a, const unsigned char **pp,
+static const char *pkey_type_to_encoder_name(int type)
+{
+    switch (type) {
+    case EVP_PKEY_RSA:
+        return "RSA";
+    case EVP_PKEY_RSA_PSS:
+        return "RSA-PSS";
+    case EVP_PKEY_DH:
+        return "DH";
+    case EVP_PKEY_DHX:
+        return "DHX";
+    case EVP_PKEY_EC:
+        return "EC";
+    case EVP_PKEY_DSA:
+    case EVP_PKEY_X25519:
+    case EVP_PKEY_ED25519:
+    case EVP_PKEY_X448:
+    case EVP_PKEY_ED448:
+        return OBJ_nid2sn(type);
+    default:
+        return NULL; /* Not supported */
+    }
+}
+
+EVP_PKEY *d2i_PrivateKey_ex(int keytype, EVP_PKEY **a, const unsigned char **pp,
                             long length, OSSL_LIB_CTX *libctx,
                             const char *propq)
+{
+    OSSL_DECODER_CTX *dctx = NULL;
+    size_t len = length;
+    EVP_PKEY *pkey = NULL;
+    EVP_PKEY **ppkey = &pkey;
+    const char *input_type = NULL;
+
+    if (keytype != EVP_PKEY_NONE) {
+        input_type = pkey_type_to_encoder_name(keytype);
+        if (input_type == NULL)
+            return NULL;
+    }
+    if (a != NULL && *a != NULL)
+        ppkey = a;
+
+    dctx = OSSL_DECODER_CTX_new_by_EVP_PKEY(ppkey, "DER", NULL, NULL,
```

Comment:
Shouldn't you actually use `input_type` here?

Reply:
Also, the documentation for `d2i_PrivateKey()` says this:

> **d2i_PrivateKey()** decodes a private key using algorithm _type_. It attempts to use any key specific format or PKCS#8 unencrypted PrivateKeyInfo format. The _type_ parameter should be a public key algorithm constant such as **EVP_PKEY_RSA**. An error occurs if the decoded key does not match _type_.  **d2i_PublicKey()** does the same for public keys.

This means that you will have to attempt this twice, once with the input structure "type-specific" and once with the input structure "pkcs8".

If you want sources of inspiration, I can offer `i2d_evp.c`